[RHCLOUD-20616] add check if ENCRYPTION_KEY is not set into InitializeEncryption

Ellen-Yi-Dong opened the pull request:
Added check for situation when ENCRYPTION_KEY is not set in InitializeEncryption()

if ENCRYPTION_KEY is empty, then send a panic error 

LINK: [RHCLOUD-20616](https://issues.redhat.com/browse/RHCLOUD-20616)

Ellen-Yi-Dong committed:
Within encryption.go, check if ENCRYPTION_KEY is set in environment in InitializeEncryption().

If ENCRYPTION_KEY is not set in environment, then panic.

Within encryption_test.go, created TestInitNoKey to test check within InitializeEncryption()

Ellen-Yi-Dong committed:
If ENCRYPTION_KEY is not set in environment when checked in encryption.go, then set to default encryption key.

Within encryption_test.go, changed TestInitNoKey in order to properly test check

petracihalova commented on util/encryption.go:
```diff
@@ -26,7 +26,9 @@ func init() {
-
+	if key == "" {
+		_ = os.Setenv("ENCRYPTION_KEY", "YWFhYWFhYWFhYWFhYWFhYQ")
+	}
```
I believe this is not working as we need

if key is empty we set there some default value - but the value is set only into environment variables, the "key" variable stays empty 
+
maybe  we can use here some nice private function to set default value, something like
```
if key == "" {
    key = setDefaultEncryptionKey()
}
```

Ellen-Yi-Dong committed:
Added helper function setDefaultEncryptionKey() within encryption.go
Added tests about helper function and to test panic error in InitializeEncryption()

Within function, checks if in local environment, encryption_key.dev file exists. If it does, then return content of file. If file does not exist, panic. If file does exist, but no symlink is created, return an error.

Within encryption_test.go, added TestSetDefaultEncryptionKey code to see if the key after being set in helper function exists in the environment.
Added TestInitEncryptionNoKey to make sure that function returns panic error when the encryption_key.dev does not exist.

Within encryption_test.go added TestInitEncryptionNoKey to check if panic error is returned when required file does not exist and the variable within the environment does not exist.

MikelAlejoBR commented on util/encryption.go:
```diff
@@ -35,6 +41,14 @@ func InitializeEncryption() {
+		return fmt.Errorf("Wrong encryption key: did not set correct default value")
+	}
+	return nil
+}
```
In this case I'd suggest not ignoring the error, and returning early. If the `Setenv` function returns an error, you know for sure that the value you wanted wasn't set.

MikelAlejoBR reviewed (commented):
I'm not sure about the use case for this one. I'm guessing this is for development purposes, so that we don't have to set an encryption key to make it work?

My feeling is that it might be better to simply let it panic as it did before, maybe with a more user friendly message. Just because:

1. We're setting a default encryption key which is public to everyone.
2. If we forget to set a proper encryption key on stage or prod —highly unlikely, I know—, I don't foresee myself in the future saying "oh, I need to check if the encryption key was properly set". Probably I would totally forget about it.

I'd like to hear everyone's views on this, please!

lindgrenj6 commented on util/encryption.go:
```diff
@@ -36,6 +44,42 @@ func InitializeEncryption() {
+	} else {
+		return nil
+	}
+	return nil
```
This else block can be removed since it's falling through if we're not on stage/prod - and the final return statement should probably be an error right? Since we want to panic and let the user know we couldn't set a default encryption key.

lindgrenj6 commented on util/encryption.go:
```diff
@@ -36,6 +44,42 @@ func InitializeEncryption() {
+		_, filename, _, ok := runtime.Caller(1)
+		if !ok {
+			return fmt.Errorf("Not possible to recover the information")
+		}
```
~I'm not following this - can you explain why we do a runtime call to see where we're at in the program?~

This would be a good example where a comment would help in the future for reference - something like 

`// fetch the filename so we know where to get the source encryption_key.dev file from`

lindgrenj6 commented on util/encryption.go:
```diff
@@ -36,6 +44,42 @@ func InitializeEncryption() {
+			if err != nil {
+				return fmt.Errorf("Unable to read file: %v", err)
+			}
+			os.Setenv("ENCRYPTION_KEY", string(body))
```
This is exactly right - the only thing I would suggest is also setting the package var `key` to `string(body)` too since we won't be calling this again (other than maybe in tests).

lindgrenj6 commented on util/encryption.go:
```diff
@@ -36,6 +44,42 @@ func InitializeEncryption() {
 }
 
+func setDefaultEncryptionKey() error {
+	if os.Getenv("SOURCES_ENV") != "stage" && os.Getenv("SOURCES_ENV") != "prod" {
```
Here you should be able to use `config.Get().Env` - we parse it at the beginning of the app startup.

lindgrenj6 commented on util/encryption.go:
```diff
@@ -36,6 +44,42 @@ func InitializeEncryption() {
+		} else if os.IsNotExist(err) {
+			return fmt.Errorf("failed to check symlink: %v", err)
+		}
+
```
Ohhhhh I see why you needed it now. 

I should have maybe been more specific. We don't want to make the symlink for the user in code - we want to just return a nice error that says "hey you need to symlink `encryption_key.dev` to `encryption_key`". 

That way the user themselves have to run `ln -s ...` we could even put the command in the error message.

lindgrenj6 reviewed (commented):
Great next pass! I think we're mostly there - we just need to let the user create the symlink rather than doing it for them.

Ellen-Yi-Dong committed:
Removing TestNoInitKey() from encryption_test.go

Deleting symlink check if the file exists and trying to pass the checks
Find correct filepath of encryption_key.dev
Clean up code

petracihalova reviewed (approved):
nice job

Ellen-Yi-Dong committed:
Merge branch 'RedHatInsights:main' into encryption_package_improvement

Ellen-Yi-Dong committed:
Remove io/outil due to deprecation

Ellen-Yi-Dong committed:
Merge branch 'RedHatInsights:main' into encryption_package_improvement
